File: 0279-perfect-squares/0279-perfect-squares.go
func numSquares(n int) int {
    cache := make(map[[1]int]int)

    var dp func(x int) int

    dp = func(x int) int {
        if x == 0 {
            return 0
        }
        if x >= 1 && x <= 3 {
            return x
        }

        if val, ok := cache[[1]int{x}]; ok {
            return val
        }

        min_sqr := int(math.Sqrt(float64(x)))
        curr := n+1


        for i:=min_sqr; i > 0; i-- {
            if x >= i*i {
                curr = min(curr, 1+dp(x-i*i))

            }
        }
        cache[[1]int{x}]=curr
        return cache[[1]int{x}]
    }
    return dp(n)

}

File: 0279-perfect-squares/0279-perfect-squares.py
class Solution:
    @cache
    def numSquares(self, n: int) -> int:
        
        if n == 0:
            return 0
        
        if 1 <= n <= 3:
            return n
        
        max_sqr = int(math.sqrt(n))
        curr = float("inf")
        for i in range(max_sqr, 0, -1):
            if n >= i**2:
                curr = min(curr, 1+self.numSquares(n-i**2))

        return curr